hierarchical PDE sampling [feature/hierarchical-sampling]

barker29 opened the pull request:
This PR is intended to implement the PDE sampler from Osborn, Vassilevski, Villa, SIAM J. Sci. Comput., 39(5), S543–S562, adapted for the finite volume context.

The new functionality is implemented in `examples/mlmc.cpp`, which can now produce coarse and fine versions of the same lognormal permeability field.

barker29 committed:
Begin work on sampler.cpp, which so far is just a copy of
finitevolume.cpp

barker29 committed:
Add W_block matrix and correlation length (inverse?) kappa. Not
at all sure this is correct, but it is boradly doing what I think
it should be doing.

barker29 committed:
Remove SPE10 coefficients, and add white noise right-hand side.

This is starting to look reasonable. Still need to have the g
scalar multiplier (see OVV (8)) and the W_\ell^{1/2} multiplier
on the right side (see unnumbered equation on bottom of p. 10)

barker29 committed:
Add --coarsening-factor command line option, clean up some
comments.

Still need to think about g scalar and W^{1/2}, but the latter
is not so important with a uniform grid, and I like what this is
doing so far.

barker29 committed:
Attempt a simple version of the scalar_g and W^{1/2} scaling.

What is this supposed to yield: similar distributions for various
correlation lengths and domain sizes?

barker29 committed:
Add a sample loop (not really used yet) and simplify Visualize() routine.

barker29 committed:
Save figures to disk and add --seed option.

barker29 committed:
Do not require SPE10 permeability file when we don't use it,
and also start looping over multiple samples.

barker29 committed:
Control solution output of samples and statistics from command line.

barker29 committed:
Also track variance in sampling.

This looks okay but is basically untested.

barker29 committed:
Cleaner output for multiple samples.

barker29 committed:
Begin JSON output for automated tests etc.

barker29 committed:
A few more things in JSON output.

barker29 committed:
Better, more meaningful output from sampler.cpp

barker29 committed:
Add basic test for PDE sampler.

barker29 committed:
Fix SPE10Problem invocation for merge.

barker29 committed:
Only upscale once for the multiple right-hand sides in sampler.cpp

barker29 committed:
Put together a rough PDESampler class, does not compile or run yet.

barker29 committed:
Compiles and runs, produces -1 times the correct answer (I think)

barker29 committed:
Remove a few vector copies (hopefully) and correctly orthogonalize
the upscaled vector for visualization / error checking.

(Can we orthogonalize on coarse grid?)

barker29 committed:
Add very basic MLMC sanity test, so at least the code runs in the test suite.

barker29 committed:
Add basic upscaling error check to sampler.cpp test.

barker29 committed:
Move PDESampler object to library (rather than sampler.cpp example)

barker29 committed:
PDESampler becomes LogPDESampler, and SimpleSampler moves into
Sampler.hpp (from mlmc.cpp)

barker29 committed:
SimpleSampler and LogPDESampler now satisfy common interface.

barker29 committed:
LogPDESampler now takes a general Upscale object instead of a
FiniteVolumeUpscale in its constructor.

barker29 committed:
Clean up from rebase/merge.

barker29 committed:
Change LogPDESampler to PDESampler, and take exponential of its
output.

Note this is not correct for the coarse sampler, where the
coefficients do not map cleanly under the nonlinear exponential
map - probably we can only do this on the constant component.

barker29 committed:
Keep track of how to represent a constant on the coarse grid,
so that we can successfully exponentiate (or take other scalar
functions) on the coarse grid.

barker29 committed:
Exponentiating on the coarse grid in PDESampler.

The pictures look okay, but a lot of the implementation is
sloppy and this is not well tested.

barker29 committed:
Clean things up just a bit; PDESampler::GetCoarseCoefficient now
appears to work as long as nothing spectral is going on.

barker29 committed:
WIP: work on sampling with PDE, some progress, examples/samper.cpp
does something reasonable visually, with longormal random field,
but a bunch of tests are failing and some of my assumptions about
how the constant vector is represented are probably wrong.

barker29 committed:
Adjust MFEM_ASSERT error checking for constant part of
interpolator.

This tolerance seems a bit high to me, but tests are passing now.

barker29 committed:
After painful checking, it looks like we need *two* Upscale
objects in mlmc.cpp (different W_blocks, among other things...)

This commit is a bit messy.

barker29 committed:
Fix use of kappa, visualize coefficient.

barker29 committed:
Take --kappa on command line.

barker29 committed:
Put copied MetisPart and CartPart code in library, instead of
copied a half dozen times in the examples.

barker29 committed:
Better names for MetisPart and CartPart.

barker29 committed:
Change fvupscale_ in PDESampler from reference to shared_ptr to
make different constructor possible.

barker29 committed:
WIP: Finish changing PDESampler constructor to build its own
Upscale object.

This is not working, mlmc.cpp is producing nonsense.

barker29 committed:
Add test for examples/mlmc with the PDESampler.

Still not sure this is working, but at least part of the issues
reported in last commit are because of unrealistically(?) small
kappa parameter - for larger kappa the coarse grid represents
things sort of reasonably.

barker29 committed:
Correct parallel seeding and add parallel test.

barker29 commented:
Don't worry about reviewing this until PR #19 is merged and reconciled with this.

gelever commented on examples/finitevolume.cpp:
```diff
@@ -180,16 +172,16 @@ int main(int argc, char* argv[])
     mfem::SparseMatrix vertex_edge = TableToMatrix(vertex_edge_table);
 
-    // Construct agglomerated topology based on METIS or Cartesion aggloemration
+    // Construct agglomerated topology based on METIS or Cartesian aggloemration
```
`aggloemration` is also misspelled, probably by me.

chakshinglee commented on examples/stest.py:
```diff
@@ -361,6 +361,23 @@ def make_tests():
+        [["./mlmc",
+          "--sampler-type", "pde",
+          "--kappa", "0.01"],
+         {"finest-p-error": 0.19865665579298755}]
```
These tests (`mlmc-pde-sampler` and `par-mlmc`) are failing in my tux box. In `mlmc.cpp`, multiple samples are generated, which error will be checked? In fact, none of the samples match this error.

gelever replied:
Same here

barker29 replied:
This test was intended to only check the last sample's error (that's where the python test scripts will look), which is not a very good test but at least tells you what I was intending.

Can you guys run this code snippet and tell me the result?
```C++
#include "../src/smoothG.hpp"

int main(int argc, char * argv[])
{
    smoothg::NormalDistribution nd0(0.0, 1.0, 0);
    std::cout << nd0.Sample() << " " << nd0.Sample() << " " << nd0.Sample() << std::endl;
    smoothg::NormalDistribution nd1(0.0, 1.0, 1);
    std::cout << nd1.Sample() << " " << nd1.Sample() << " " << nd1.Sample() << std::endl;
    return 0;
}
```
On my machine the result is
```
0.0708592 0.302805 1.12279
1.58275 -1.40287 -0.549746
```
Did we have a similar issue with the random graph generation? How did we make sure that was consistent across platforms?

gelever replied:
output on my machine:
```
./examples/testss 
1.12279 0.302805 0.0708592
-0.549746 -1.40287 1.58275
```
It might be changing how `Sample()` is executed, [left to right?](https://en.cppreference.com/w/cpp/language/eval_order)

Although this shouldn't be an issue is this PR, it just calls `Sample()` when setting a vector.

gelever replied:
Another test:
``` c++
#include "../src/smoothG.hpp"

int main(int argc, char * argv[])
{
    smoothg::NormalDistribution nd0(0.0, 1.0, 0);
    double s1 = nd0.Sample();
    double s2 = nd0.Sample();
    double s3 = nd0.Sample();
    std::cout << s1 << " " << s2 << " " << s3 << std::endl;

    smoothg::NormalDistribution nd1(0.0, 1.0, 1);
    double s4 = nd1.Sample();
    double s5 = nd1.Sample();
    double s6 = nd1.Sample();
    std::cout << s4 << " " << s5 << " " << s6 << std::endl;

    return 0;
}
```
outputs:
```
1.12279 0.302805 0.0708592
-0.549746 -1.40287 1.58275
```

gelever replied:
When i run `mpirun -np 2 ./mlmc --sampler-type pde --kappa 0.01 --perm fake_perm.txt`, I get the same results as the test.  I'm guessing there is an issue reading the file and different permeabilities are being used.

If the user specified file is not found, that should actually throw/error (vs silently using `BlankPermeability`)

barker29 replied:
Thanks @gelever , you're right, and I don't think it has anything to do with the random numbers. Hopefully fixed.

chakshinglee replied:
@barker29 those tests now all pass on my tux box.

chakshinglee commented on examples/sampler.cpp:
```diff
@@ -0,0 +1,405 @@
+        for (int i = 0; i < sol_upscaled.Size(); ++i)
+            sol_upscaled(i) = std::log(sol_upscaled(i));
+        fvupscale->Orthogonalize(sol_upscaled); // can we orthogonalize on coarse grid?
+        int coarse_iterations = fvupscale->GetCoarseSolveIters();
```
https://github.com/LLNL/smoothG/blob/ff065cbdbc512c92b44162ad6daed04fcbf5da81/src/MinresBlockSolver.cpp#L218

https://github.com/LLNL/smoothG/blob/ff065cbdbc512c92b44162ad6daed04fcbf5da81/src/MinresBlockSolver.cpp#L236

We haven't changed the setting in `MinresBlockSovler`, so the number of iteration and solve time will accumulate every time when the solver is called to solve a problem.

chakshinglee commented on examples/sampler.cpp:
```diff
@@ -0,0 +1,405 @@
+        auto sol_upscaled = fvupscale->Interpolate(sol_coarse);
+        for (int i = 0; i < sol_upscaled.Size(); ++i)
+            sol_upscaled(i) = std::log(sol_upscaled(i));
+        fvupscale->Orthogonalize(sol_upscaled); // can we orthogonalize on coarse grid?
```
I think you can do so once you have the coarse representation of constant 1 (which you have made it available in this PR). 

In fine level this is what we do:
vec -= (1^T * vec) 1

Let P be the interpolation matrix. If we have a coarse solution vec_c, and a coarse representation of fine constant 1 (denoted by 1_c = P 1), then of course we can interpolate all the coarse quantities to fine level and use the fine level orthogonalization (and this is what the above few lines is doing):
P vec_c -= ( (P 1_c)^T * (P vec_c) ) P 1_c

The above formula can be simplified because our P satisfies P^T P = I. So 
( (P 1_c)^T * (P vec_c) ) P 1_c = ( (1_c)^T P^T * P vec_c) P 1_c = P ( (1_c)^T * vec_c) 1_c

So the coarse level orthogonalization is 
vec_c -= ( (1_c)^T *  vec_c) 1_c

You can interpolate to fine level afterwards if you want.

gelever commented:
`src/Sampler.hpp` needs `#include <random>` on my machine to build.  Otherwise, there are compile errors about not finding the random generator stuff.

barker29 committed:
Merge remote-tracking branch 'origin/master' into feature/hierarchical-sampling

barker29 committed:
Correct spelling.

barker29 committed:
MinresBlockSolver now sets rathan adds iterations, timing, to be consistent with HybridSolver.

barker29 committed:
This commit actually does what the last commit message claimed to do.

barker29 committed:
Add <random> include to Sampler.hpp

barker29 committed:
Add SPE10Problem constructor that allows blank (constant)
permeability in a nicer way with better error-handling for reading
actual permeability data.

barker29 committed:
mlmc.cpp now dumps gridfunctions for later postprocessing into pictures.

chakshinglee commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,233 @@
+    mfem::SparseMatrix W_block = SparseIdentity(vertex_edge.Height());
+    W_block *= cell_volume_ * kappa * kappa;
+    mfem::Vector one_weight(vertex_edge.Width());
+    one_weight = 1.0;
```
If "kappa^2 u - \triangle u = g" is the pde considering here, the weight will need to take geometry (like face area, characteristic length) into account?

barker29 replied:
Yes. The right thing to do is to use `FiniteVolumeMassIntegrator`, but that requires a BilinearForm and a Mesh, and that seems like a lot of machinery to bring into this class, and I am lazy. But probably I should just do it, since this class only makes sense in the finite volume context anyway.

chakshinglee replied:
Actually, the current method that you implement (if we exclude cell volume and boundary condition as well) can be interpreted as a graph-based sampler through graph Laplacian. Clearly it has its PDE origin in our case, but this extension may be used to apply to other problems (if we eventually encounter one some day Zzzz). 

Cell volume and other geometrical factors can be integrated into the graph through vertex and edge weights. So that seems to be not a big problem.

For boundary condition, apparently it has been an issue in the original PDE-based method, so it may be not too bad to use Neumann condition, which means simply ignore the boundary. In the reference paper, they solve problem on an extended domain, not sure how to do that in graph nicely if we remove the notion of boundary (the equivalent thing should be to only extend the graph at the vertices that is on the "boundary").

barker29 commented:
@chakshinglee do the two solvers give different results even for the first sample? Or just on later samples?

chakshinglee commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,233 @@
+        fvupscale_->Restrict(rhs_fine_, rhs_coarse_);
+    mfem::Vector coarse_sol = fvupscale_->GetCoarseVector();
+    fvupscale_->SolveCoarse(rhs_coarse_, coarse_sol);
+    coarse_sol *= -1.0; // ?
```
Compare with `GetFineCoefficient`, the solution here is not orthogonalized to constants. I dont know if that is necessary, I just notice the difference between the two (and we have the tool to do the average).

chakshinglee commented on src/GraphCoarsen.cpp:
```diff
@@ -94,6 +98,25 @@ void GraphCoarsen::BuildPVertices(
+                    const double cval = coarse_constant_rep_(coarse_vertex_dof_counter);
+                    if (cval == 0.0)
+                    {
+                        // @todo this should probably depend on volume size etc?
```
It may not depend on the volume of elements in fine level (unless the basis for the fine level is not constant 1 on an element), but it will depend on the volume of aggregate, which I guess is why you take `1.0 / target_i(j, k)` here.

An alternative way to compute the representation of constant vector in coarse level is to make use the fact that `Pvertices` is orthogonal, so

constant_one_coarse = Pvertices * constant_one_fine 

Then the volume factor will be taken care of automatically, and it should be extensible to multilevel as long as we construct `Pvertices` properly.

barker29 replied:
I feel fairly dumb for not realizing this myself.

chakshinglee reviewed (approved):
@barker29 This PR looks good to me, in particular, I think the PDE sampler is very cool. There is only one thing I want to change, which is due to my mistake: 
- [ ] Enforce update of assembled fine M in both solver options so that printed errors are consistent.

There are other suggestions in the individual comments, some of them I am not sure if it is necessary for illustration purpose (e.g. taking average zero in the coarse coefficient generation, it makes sense to have similar settings in both levels though).

gelever commented on examples/mlmc.cpp:
```diff
@@ -160,6 +143,19 @@ int main(int argc, char* argv[])
+    int num_samples = 3;
+    args.AddOption(&num_samples, "--num-samples", "--num-samples",
+                   "Number of samples to draw and simulate.");
+    bool blank_perm = false;
```
would it be simpler to have the default empty filename string imply blank permutation?  That way all the complexity is moved into `IPF::ReadPermeabilityFile`. Everything else would be unchanged: there would be the single spe10 constructor the and the spe10 object would not need to be a unique pointer.

gelever commented on examples/mlmc.cpp:
```diff
@@ -160,6 +143,19 @@ int main(int argc, char* argv[])
+                   "Instead of loading SPE10 data, use constant permeability.");
     args.Parse();
     if (!args.Good())
     {
```
`MPI_Finalize();` is incorrect in this block.  The `mpi_session` from above takes care of it.

edit: Same w/ `examples/sampler.cpp`. Examples handle MPI inconsistently.  We should probably use `mpi_session` in all the examples and avoid manually `MPI_Finalize()`.

gelever commented on src/Sampler.hpp:
```diff
@@ -0,0 +1,182 @@
+    virtual void NewSample() {}
+
+    /// return current sample realized on fine mesh
+    virtual mfem::Vector& GetFineCoefficient() = 0;
```
This can return `const mfem::Vector&`.

gelever commented on src/Upscale.hpp:
```diff
@@ -123,6 +123,12 @@ class Upscale : public mfem::Operator
     virtual MixedMatrix& GetCoarseMatrix();
     virtual const MixedMatrix& GetCoarseMatrix() const;
 
+    virtual const mfem::SparseMatrix& GetVertexProjection() const { return coarsener_->get_Pu(); }
```
Is this used anywhere?

barker29 replied:
No. But I could swear I put it in there for some reason. Gone now.

gelever commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,233 @@
+    return coefficient_coarse_;
+}
+
+/// @todo ugly hack for debugging, @deprecated
```
Adding an already deprecated function in a PR doesn't sound like a good idea.
If it is still needed, maybe remove the tag.

barker29 replied:
I removed the tag. I like what this does in `examples/sampler.cpp`, so I think it is worth keeping around.

gelever commented on src/utilities.hpp:
```diff
@@ -421,6 +422,10 @@ void RescaleVector(const mfem::Vector& scaling, mfem::Vector& vec);
 */
 void GetElementColoring(mfem::Array<int>& colors, const mfem::SparseMatrix& el_el);
 
+void FVMeshCartesianPartition(
```
I know this was just copied from the earlier example files, but some of these can be made const.  We might as well take care of it in this PR.

gelever commented on .dir-locals.el:
```diff
@@ -9,5 +9,7 @@
          ((c-basic-offset . 4)
           (c-indent-level . 4)
-          (c-file-style . "BSD"))))
+          (c-file-style . "BSD")))
```
Does this file even belong in the project?

gelever commented on src/Upscale.hpp:
```diff
@@ -123,6 +123,12 @@ class Upscale : public mfem::Operator
+
+    /// @todo I feel like needing this reflects some leaky abstractions
+    ///       or bad design in one way or another.
+    virtual const GraphCoarsen& GetGraphCoarsen() const { return coarsener_->get_GraphCoarsen_ref(); }
```
Exposing this can be avoided using @chakshinglee's note on computing coarse constant using `P_vertex`.  Either in this class or outside utility function.
```
mfem::Vector GetCoarseConstant(const Upscale& upscale) const
{
         mfem::Vector ones = upscale.GetFineVector();
         ones = 1.0;

         return upscale.Restrict(ones);
}

barker29 committed:
Remove unnecessary MPI_Finalize() calls in mlmc.cpp.

gelever commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,233 @@
+        nu_parameter = 0.5;
+    double ddim = static_cast<double>(dimension);
+    scalar_g_ = std::pow(4.0 * M_PI, ddim / 4.0) * std::pow(kappa, nu_parameter) *
+                std::sqrt( tgamma(nu_parameter + ddim / 2.0) / tgamma(nu_parameter) );
```
`tgamma` should probably be `std::tgamma` to match `pow` and `sqrt`

barker29 committed:
A few added const qualifiers.

gelever commented on examples/sampler.cpp:
```diff
@@ -0,0 +1,405 @@
+    }
+
+    if (myid == 0)
+        std::cout << picojson::value(serialize).serialize() << std::endl;
```
can this output w/ `picojson::value(serialize).serialize(true)` instead? It is easier to read when the output is vertically stacked.

barker29 committed:
FiniteVolumeMLMC now consistently updates the fine M matrix for
both solvers.

barker29 committed:
Remove .dir-locals.el from the git repo.

barker29 committed:
Remove MPI_Finalize() and prettier JSON printing in sampler.cpp

barker29 committed:
Add OrthogonalizeCoarse() and use it by default in SolveCoarse()

barker29 committed:
Simpler and more general way of constructing coarse_constant_rep_

barker29 committed:
Even better way of building coarse_constant_rep.

gelever commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,232 @@
+    for (int i = 0; i < coefficient_coarse_.Size(); ++i)
+    {
+        coefficient_coarse_(i) =
+            std::exp(coefficient_coarse_(i) / coarse_constant_rep(i)) * coarse_constant_rep(i);
```
This breaks if using more than just the constant basis vectors yeah?  Because `coarse_constant_rep[i]` is zero in some places and consequently divide by zero.
For example, `./sampler -m 4 -t 1.0 -vis` fails to run.

gelever commented on examples/sampler.cpp:
```diff
@@ -0,0 +1,404 @@
+        SaveFigure(mean_upscaled, ufespace, coarsenamev.str());
+        std::stringstream finenamev;
+        finenamev << "fine_variance";
+        SaveFigure(mean_fine, ufespace, finenamev.str());
```
`mean_upscaled` and `mean_fine` are both saved twice.  Is `coarse_variance` supposed to save `m2_upscaled` (and fine equivalent)?

These can be done a bit cleaner w/o stringstream
```c++
{
        SaveFigure(mean_upscaled, ufespace, "coarse_mean");
        SaveFigure(mean_fine, ufespace, "fine_mean");
        SaveFigure(m2_upscaled, ufespace, "coarse_variance");
        SaveFigure(m2_fine, ufespace, "fine_variance");
}
```

gelever commented on src/Sampler.cpp:
```diff
@@ -0,0 +1,232 @@
+        fvupscale_->Restrict(rhs_fine_, rhs_coarse_);
+    coefficient_coarse_.SetSize(rhs_coarse_.Size());
+    fvupscale_->SolveCoarse(rhs_coarse_, coefficient_coarse_);
+    coefficient_coarse_ *= -1.0; // ??
```
I think this `*= -1.0;` needs to go into `Upscale::SolveCoarse` instead.  There may not  be anywhere else where it was used.

gelever commented on src/Upscale.cpp:
```diff
@@ -104,6 +104,7 @@ void Upscale::SolveCoarse(const mfem::BlockVector& x, mfem::BlockVector& y) cons
 
     coarse_solver_->Solve(x, y);
     y *= -1.0;
+    OrthogonalizeCoarse(y);
```
This (lines 106 and 107) should probably be added to the non-block `SolveCoarse` also (right after line 90).

gelever commented on src/Upscale.cpp:
```diff
@@ -244,6 +245,20 @@ void Upscale::Orthogonalize(mfem::BlockVector& vect) const
+    double local_dot = (vect * coarse_constant_rep);
+    double global_dot;
+    MPI_Allreduce(&local_dot, &global_dot, 1, MPI_DOUBLE, MPI_SUM, comm_);
+    vect.Add(-global_dot, coarse_constant_rep);
```
@chakshinglee Referring to your note from an earlier commit, should this include the fine level global size somewhere to compute average?  Or is that automatically handled already?

`vect.Add(-global_dot / fine_global_size, coarse_constant_rep);`

chakshinglee replied:
Good catch, you are right. I missed the inner product of constant vector. I actually missed that in the fine level formula, and consequently the coarse one as well (because it was derived from fine). It should be v -= (v * 1 ) / (1 * 1) * 1. Similarly for coarse level v_c -= (v_c * 1_c ) / (1_c * 1_c) * 1_c.

The inner product (1_c * 1_c) should have the same value as (1 * 1), which equals `fine_global_size`. Either include that factor in the normalization, or if we store this constant representation, we can just store the normalized version, i.e. store 1_c / || 1_c ||, then (1_c * 1_c) = 1 and we dont need that factor.

barker29 committed:
Fix divide-by-zero in GetCoarseCoefficientForVisualization()

barker29 committed:
Save the correct figures, and simplify the figure-saving code.

barker29 committed:
Move scaling by -1 from PDESampler::GetCoarseCoefficient to
Upscale::SolveCoarse(), where it seems to be more consistent.

I am a bit confused about what is going on here.

barker29 committed:
Better / more systematic / normalized orthogonalization on coarse
grid.

(Also slight rebaselining of MLMC tests because errors are
slightly better with this orthogonalization than they were
before.)

barker29 commented:
A lot of comments here, so I want to write down what I think still needs to be done so you guys can check if I missed something:

- [ ] PDESampler should handle real finite volumes, or at least document how to use the graph structure to do a finite volume problem.
- [ ] Save the `coarse_one_rep_` vector instead of generating it every time.
- [ ] Better handling of empty filename / blank permeability file in `mlmc.cpp`

Am I missing anything?

chakshinglee commented:
Just found a minor bug that is not from this PR, but maybe it can be corrected in this PR instead of open another one.

https://github.com/LLNL/smoothG/blob/b1b60725b4afc39b027b74d2734939a77612a4ab/src/HybridSolver.cpp#L906-L911

Line 908 should call `SetRelTol` instead of `SetMaxIter`.

barker29 committed:
Bug fix: HybridSolver::SetRelTol()

barker29 committed:
--seed on command line, and also save log(coefficient) to file.

barker29 committed:
Saner handling of blank default permeability in mlmc.cpp.

barker29 committed:
Save coarse_constant_rep_ in Upscale object for repeated access.

barker29 committed:
Pass edge weights to PDESampler constructor so that its internal
upscaling object better reflects the finite volume context from
the caller in mlmc.cpp
